dimod/serialization/lp.py

```diff
@@ -0,0 +1,268 @@
+    grammar.ignore(comment_style)
+
+    return grammar
+
+
+def remove_strings(string, strings_to_remove):
+    """Replaces an iterable of strings in removables
+        if removables is a string, each character is removed """
+    for r in strings_to_remove:
+        try:
+            string = string.replace(r, "")
+        except TypeError:
+            raise TypeError("Strings_to_remove contains a non-string element")
+    return string
+
+
+def get_variables(parse_output: ParseResults,
+                  lower_bound_default: typing.Optional[int] = None,
+                  upper_bound_default: typing.Optional[int] = None) -> Dict:
+    """
+    Returns set of variables and dict of variable types and bounds
+    Args:
+        parse_output: the parse results encoding the LP file
+        lower_bound_default: the lower bound of the integer variables, in case they are not specified
+        upper_bound_default: the upper bound of the integer variables in case they are not specified
+
+    Returns:
+        a dictionary for each variable containing type, upper and lower bounds
+    """
+    # handle the default bounds
+    if lower_bound_default is None:
+        lb_def = -infinity
+    else:
+        lb_def = lower_bound_default
+
+    if upper_bound_default is None:
+        ub_def = infinity
+    else:
+        ub_def = upper_bound_default
+
+    all_vars = set()
+    # default variable type for LP file
+    # should be continuous, even though
+    # they are not supported in CQMs
+    variables_info = defaultdict(lambda: {"type": "c", "lb": lb_def, "ub": ub_def})
+
+    # scan the objective
+    for oe in parse_output.objective:
+        if isinstance(oe, str):
+            continue
+
+        else:
+            if len(oe) == 2:
+                if oe.name != "":
+                    all_vars.add(oe.name[0])
+                else:
+                    # not supported yet.. although they can be replaced by CQM function
+                    raise ValueError("pure quadratic terms are not supported, found {}^2".format(oe.squared_name[0]))
+            elif len(oe) == 3:
+                all_vars.update([oe.name[0], oe.second_var_name[0]])
+
+    # scan the constraints
+    for c in parse_output.constraints:
+
+        # scan linear terms
+        if c.lin_expr:
+            all_vars.update([le.name[0] for le in c.lin_expr])
+
+        # scan quadratic terms of constraints
+        if c.quad_expr:
+            for qe in c.quad_expr:
+                if len(qe) == 3:
+                    all_vars.update([qe.name[0] for qe in c.quad_expr])
+                    all_vars.update([qe.second_var_name[0] for qe in c.quad_expr])
+                elif len(qe) == 2:
+                    # not supported yet
+                    raise ValueError("pure quadratic terms are not supported, found {}^2".format(qe.squared_name[0]))
+
+    # check the binary variables:
+    for n in parse_output.binaries:
+
+        variables_info[n]["ub"] = 1
```

Comment:
again, I think we could just allow this to be `None` and then allow the later `add_variable` method(s) to infer them from the vartype.